src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java

```diff
@@ -0,0 +1,134 @@
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.persistence.FileRotate;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.LineIterator;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * PersistableSlidingWindow is a SlidingWindow which can have its data written to and read from disk
+ *
+ * <p>The sliding window retains data for slidingWindowSizeInSeconds and aggregates values into time
+ * buckets of width bucketSizeInSeconds. For example if slidingWindowSizeInSeconds is 86400 (1 day)
+ * and bucketSizeInSeconds is 3600 (1 hr) then there will be at most 25 data points in the sliding window
+ * at any given time. It's 25 and not 24 because the oldest datapoints are removed lazily based on
+ * arrival time. This shouldn't impact most consumers of this data structure.
+ *
+ * <p>The data contained in this sliding window is effectively a time series. Consumers may extend this
+ * class and define more complex aggregations or analytics functions.
+ */
+public class PersistableSlidingWindow extends SlidingWindow<SlidingWindowData> {
+  private static final Logger LOG = LogManager.getLogger(PersistableSlidingWindow.class);
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  protected static final String SEPARATOR = "\n";
+
+  private Path pathToFile;
+
+  protected SlidingWindowData aggregatedData;
+  protected final double bucketSizeMs;
+
+  public PersistableSlidingWindow(int slidingWindowSize,
+                                  int bucketSize,
+                                  TimeUnit timeUnit,
+                                  Path filePath) {
+    super(slidingWindowSize, timeUnit);
+    this.bucketSizeMs = timeUnit.toMillis(bucketSize);
+    this.pathToFile = filePath;
+    try {
+      if (Files.exists(pathToFile)) {
+        loadFromFile(pathToFile);
+      } else {
+        Files.createFile(pathToFile);
+      }
+    } catch (IOException e) {
+      LOG.error("Couldn't create file {} to perform young generation tuning", pathToFile, e);
+      throw new IllegalArgumentException("Couldn't create or read a file at " + filePath);
+    }
+  }
+
+  @Override
+  public void next(SlidingWindowData slidingWindowData) {
+    if (aggregatedData == null) {
+      aggregatedData = new SlidingWindowData(slidingWindowData.getTimeStamp(), slidingWindowData.getValue());
+      return;
+    }
+    long timestampDiff = slidingWindowData.getTimeStamp() - aggregatedData.getTimeStamp();
+    if (timeUnit.convert(timestampDiff, TimeUnit.SECONDS) > bucketSizeMs) {
```

Comment:
Is this functionally correct? IIUC, this expects `timestampDiff` to be in `SECONDS`, converts it to `timeUnit` and compares it against `bucketSizeMs` which is in milliseconds.
Is `timeUnit` guaranteed to be in ms? Is `timestampDiff` guaranteed to be in seconds?

Reply:
Good catch, fixed and added a UT to catch this.

SlidingWindowData timestamps must be in ms (SlidingWindow assumes this). The bucket length and overall window size can be specified using any timeunit.